### DISCUSSION TITLE
Cluster certificate best practices

### DISCUSSION THREAD

**Discussion by krivulcik** [ravendb/ravendb#14474] [Q&A] [answered]
We have two clusters with the issue at which we took a look with Oren yesterday: When I open the studio on a non-leader node and issue a command that can be handled only by the leader node, the local node tries to redirect/forward the command to the leader node, but fails to authenticate itself against the leader node with a client certificate.

See https://github.com/ravendb/ravendb/discussions/14454#discussioncomment-3049751 for further context.

This issue persists even after configuring the nodes to communicate over the private internal Azure network. Configuring the `PublicServerUrl.Tcp.Cluster` value in settings.json helped with the cluster instability we were investigating in the original discussion: the cluster is now almost completely stable, and no voting occurs. However, the issue with the non-leader node not authenticating correctly against the leader node persists.

I believe that a networking issue can be safely ruled out. I manually verified using netcat that existing connections are kept alive for a long time (1 hour of inactivity). This is not the case when using public IP addresses - the connections are dropped after 4 minutes (this seems to be an Azure-specific thing).

When I try to access another node in a browser, the browser doesn't offer a certificate to select. This implies that the certificates are not installed in the server certificate stores.

When I try to access another node using the certificate configured in `Security.Certificate.Path` in settings.json, the call succeeds:

```
> $clientCert = Get-PfxCertificate -FilePath C:\RavenDB\Server\cluster.server.certificate.clustername.pfx

> Invoke-WebRequest -Method Get https://[OTHER NODE URL]/admin/debug/node/ping -Certificate $clientCert


StatusCode        : 200
StatusDescription : OK
Content           : {"Result":[{"Url":"https://.....ravendb.community","TcpInfoTime":3,"SendTime":7,"ReceiveTime":8,"Er
                    ror":null,"Log":["Received tcpInfo: \r\nurls: tcp://10.1.0.8:38888,
                    tcp://.....ravendb.community:3888...
RawContent        : HTTP/1.1 200 OK
                    Transfer-Encoding: chunked
                    Raven-Server-Version: 5.3.103.53023
                    Content-Type: application/json; charset=utf-8
                    Date: Fri, 01 Jul 2022 13:53:25 GMT
                    Server: Kestrel
```

This implies that the cluster certificate can be used for HTTPS requests to other nodes, when used.

From this, it seems that the local RavenDB doesn't use the correct certificate (or any certificate?) for the forwarded/redirected requests to the leader node.

Is there any recommendation about the certificate management I might have missed during the cluster setup?

Should the cluster certificate be installed in the system store? (Would it need to be updated when it gets renewed from letsencrypt?)

The error when issuing commands on a non-leader node is the only manifestation of this issue. The cluster works okay in every other way as far as we can evaluate. Replication, failover, load balancing, etc., - everything works correctly.

**Comment by ayende** [comment, marked as answer]
You have _something_ in the middle that intercept the requests.

RavenDB doesn't use the local store for the client certificate it has, only for validating root CA, etc.

For _all_ requests by RavenDB, the server certificate (the same one we use to server requests ourselves) will be used to auth as the client certificate.

Your scenario looks like something is intercepting _some_ of the traffic.

**Reply by krivulcik** [reply]
I just created a cluster on my laptop where three instances of RavenDB run next to each other, secured by LetsEncrypt certificates.

The issue is present even in this environment where there is definitely no traffic manipulation.

`settings.json` files for the nodes look like this:

Node A:
```
{
  "License.Eula.Accepted": true,
  "Security.Certificate.LetsEncrypt.Email": "###",
  "Setup.Mode": "LetsEncrypt",
  "Security.Certificate.Path": "###\\RavenDB-5.3.103-windows-x64-A\\cluster.server.certificate.clustertest.pfx",
  "ServerUrl": "https://0.0.0.0:10301",
  "ServerUrl.Tcp": "tcp://0.0.0.0:10311",
  "ExternalIp": "###",
  "PublicServerUrl": "https://a.###.development.run:10301",
  "PublicServerUrl.Tcp": "tcp://a.###.development.run:10311",
  "DataDir": "RavenData",
  "PublicServerUrl.Tcp.Cluster": "tcp://127.0.0.1:10311"
}
```

Node B:
```
{
  "License.Eula.Accepted": true,
  "Security.Certificate.LetsEncrypt.Email": "###",
  "Setup.Mode": "LetsEncrypt",
  "Security.Certificate.Path": "###\\RavenDB-5.3.103-windows-x64-A\\cluster.server.certificate.clustertest.pfx",
  "ServerUrl": "https://0.0.0.0:10302",
  "ServerUrl.Tcp": "tcp://0.0.0.0:10312",
  "ExternalIp": "###",
  "PublicServerUrl": "https://b.###.development.run:10302",
  "PublicServerUrl.Tcp": "tcp://b.###.development.run:10312",
  "DataDir": "RavenData",
  "PublicServerUrl.Tcp.Cluster": "tcp://127.0.0.1:10312"
}
```

Node C:
```
{
  "License.Eula.Accepted": true,
  "Security.Certificate.LetsEncrypt.Email": "###",
  "Setup.Mode": "LetsEncrypt",
  "Security.Certificate.Path": "###\\RavenDB-5.3.103-windows-x64-A\\cluster.server.certificate.clustertest.pfx",
  "ServerUrl": "https://0.0.0.0:10303",
  "ServerUrl.Tcp": "tcp://0.0.0.0:10313",
  "ExternalIp": "###",
  "PublicServerUrl": "https://c.###.development.run:10303",
  "PublicServerUrl.Tcp": "tcp://c.###.development.run:10313",
  "DataDir": "RavenData",
  "PublicServerUrl.Tcp.Cluster": "tcp://127.0.0.1:10313"
}
```

The cluster view is the following:
![image](https://user-images.githubusercontent.com/7997055/177124526-8f062c53-88be-4e33-81ab-49b4948ade06.png)

An attempt to demote node A from node C (node B is the leader) yields the following:
![image](https://user-images.githubusercontent.com/7997055/177124798-a9f3a6e7-0d28-4af6-8122-2e97a3b63847.png)

This is easily reproducible with just a clean local cluster setup. Is there anything I may have missed to make this work? Any suggestions about how to investigate further?

**Reply by ayende** [reply]
Can we set a call to look into that? That shouldn't be happening.

**Reply by krivulcik** [reply]
Turns out to be browser issue (Firefox): https://issues.hibernatingrhinos.com/issue/RavenDB-18901
